src/Compilers/Core/Portable/PEWriter/PeWriter.cs

```diff
@@ -64,16 +106,13 @@ internal static class PeWriter
             // based on the contents of the generated stream.
             Debug.Assert(properties.PersistentIdentifier == default(Guid));
 
-            var ilBuilder = new BlobBuilder(32 * 1024);
-            var mappedFieldDataBuilder = new BlobBuilder();
-            var managedResourceBuilder = new BlobBuilder(1024);
-
+            var emitBuilders = new EmitBuilders();
```

Comment:
@AlekseyTs, @333fred 

Last night I was considering going ahead and fixing the [LinkSuffix / LinkPrefix bug][blobbuilder-bug] and realized my current approach is flawed. Once that bug is fixed and a new version of SRM is used with Roslyn then my code will start throwing cause `Count == 0` builders will be double freed. This will very concretely happen in Source Build.

After some thought I decided a more durable fix is to pass `null` into `ExtendedPEBuilder` for these two `BlobBuilder` instead of an empty `BlobBuilder`. The `null` is an allowed case and is not impacted by the bug fix. There were two ways to achieve this:

1. Add a simple check after the `BuildAndMetadataAndIL` and free + `null` out the fields if their Count is zero. This is a smaller change but felt a bit off. Like I was fixing the symptom not the cause. 
2. Do not allocate the `BlobBuilder` unless we're going to put content into it. This change produces bigger code churn but feels more correct.

Ended up going with approach (2) here. If you feel churn is too big can use approach (1) instead.